advanced-game-design-team/Artifacts/Raycast Collisions/README.md

```diff
@@ -0,0 +1,112 @@
+# Raycasted Point Collision
+
+![](./rays.gif)
+
+## Introduction
+
+This guide explores a 2D raycasting system designed for detecting collisions between rays emanating from a point (typically the mouse cursor) and rectangular obstacles. The system comprises several key components that work together to create an efficient collision detection mechanism.
+
+## Core Components
+
+### 1. Ray Generation
+
+The system generates rays from the mouse cursor position in a 360-degree circle. The number of rays is defined by `NUM_RAYS`, which is set to 360 in this implementation, effectively creating one ray per degree.
+
+```cpp
+struct ray {
+    point_2d origin;
+    vector_2d direction;
+};
+
+// In the main loop:
+point_2d mouse = mouse_position();
+std::vector<ray_hit> hits = cast_rays(obstacles, mouse, NUM_RAYS, MAX_DISTANCE);
+```
+
+### 2. Obstacle Generation
+
+Rectangular obstacles are randomly generated at the start of the program, creating a dynamic environment for ray collision:
+
+```cpp
+std::vector<rectangle> generate_random_obstacles(int num_obstacles) {
+    std::vector<rectangle> obstacles;
+    for (int i = 0; i < num_obstacles; ++i) {
+        double x = rnd(SCREEN_WIDTH);
+        double y = rnd(SCREEN_HEIGHT);
+        double width = rnd(10, 300);
+        double height = rnd(10, 20);
+        obstacles.push_back(rectangle_from(x, y, width, height));
+    }
+    return obstacles;
+}
+```
+
+### 3. Ray-Rectangle Intersection Detection
+
+The core of the collision detection is the `ray_rectangle_intersection` function, which implements the "slab" method for Ray-AABB (Axis-Aligned Bounding Box) intersection:
```

Comment:
I'd hazard that far less than half of students coming into capstone will have encountered AABB collision before, so I think incorporating a brief, very high-level explanation of what it is would be helpful here (in general, and/or contextualized to this specific example), ideally with a pre-screened Oscar-Approved™ link for further reading for those who want it.

e.g.:
The core of the collision detection is the `ray_rectangle_intersection` function, which uses AABB (Axis-Aligned Bounding Box) collision detection. This is a specific way of checking for collisions between objects - in our case between rays and the rectangle obstacles we created earlier. Here we've implemented the "slab" method for Ray-AABB intersection: